Add invitation management

grokdesigns opened the pull request:
## Community Contribution License Agreement
By creating this pull request, I grant the project maintainers an unlimited,
perpetual license to use, modify, and redistribute these contributions under any terms they
choose, including both the AGPLv3 and the Fossorial Commercial license terms. I
represent that I have the right to grant this license for all contributed content.

## Description
![Screen Recording 2025-04-08 at 18 16 12](https://github.com/user-attachments/assets/ae923a67-cd70-4e01-b43d-83117a3ec63f)

Was looking for a way to manage/revoke/etc. invitations and didn't see anything, so I hope that functionality isn't just somewhere I didn't notice! Tested with multiple orgs/invites. Confirmed revoked invitations give user error message when using invite link. Considered moving "Create User" button to invitations page, but probably too much of a change to base functionality. Could duplicate "Create User" on both pages if desired.

## How to test?
Build from source and update Docker image.

grokdesigns commented:
![image](https://github.com/user-attachments/assets/2287c63e-10e5-4ced-a562-bb7f6e14029a)
![image](https://github.com/user-attachments/assets/121a757e-9861-45af-b586-1d4d8216e765)

Latest commit adds child menu item support to the SidebarNav function function, moved Invitations under Users and it will only appear if there are items in the userInvites table.

miloschwartz commented:
Hey thanks for working on this. This is one of those things that we cut down on at the beginning to release faster but we've always wanted. I will review this when I have some free time soon- hopefully this weekend or earlier.

grokdesigns commented:
@miloschwartz Glad to help! Quite a few changes in this most recent commit.

- Changed icon to more universal down-right arrow to indicate Invitations is a sub-menu of Users. Still only visible if invitation exist.
- Added the regenerate invitation functionality and have tested it with all combinations I can think of. This uses a modal popup like the Invite User function does, but does not allow changing the email address or user type, just the number of days the invitation is valid so it can be extended. Most of this is in the new RegenerateInvitationForm.tsx file.
- Replaced the method for keeping track of how many invites have been sent per hour in inviteUser.ts. Took the logic out of the initial Invite User functionality. The initial invite will only let you create one invitation for a given email address. Moved the new inviteTracker logic to the regeneration of invites. User will be able to create initial invite and then can regenerate up to 3 times within a one hour period. I changed your exiting logic to use node-cache instead of the previous solution where there was a comment of "// TODO: this could cause memory increase over time if the object is never deleted". I think node-cache will solve that, as it should clean up after itself. I haven't actually tested to ensure it is successfully reset after an hour, so should test that.
- Changed Invite User and Regenerate Invitation modals so that they always default to send email. Before, if you invited multiple users or regenerated multiple invitations, it would maintain the last state it was in. I think setting the default every time the modal opens is a cleaner user experience.
- Changed invitation email subject from "You're invited to join a Fossorial organization" to "You're invited to join {orgId}". Went back and forth on this one, but I think Fossorial might be confusing to end/less-technical users. Maybe this customization could be behind a subscription key?

Screenshots:
![image](https://github.com/user-attachments/assets/c879a22c-ddf2-4480-9071-cfb61b8ebdeb)
![image](https://github.com/user-attachments/assets/7ad6c748-8030-4669-b0cb-3a0731c85a08)
![screenshot_20250409_20 51 53_32ibIt](https://github.com/user-attachments/assets/ff2caa16-18ae-4518-9a85-ed8cad526c7e)
![screenshot_20250409_20 53 13_xqEtbD](https://github.com/user-attachments/assets/ab1b3632-470a-4ffa-bc5b-e68fd6916897)
![screenshot_20250409_20 50 02_gTZVPf](https://github.com/user-attachments/assets/066a8e12-60a2-4c38-869f-87f260325af7)

grokdesigns commented:
@miloschwartz Glad to help! Quite a few changes in this most recent commit.

- Changed icon to more recognizable down-right arrow to indicate Invitations is a sub-menu of Users. Still only visible if invitations exist.

- Added the regenerate invitation functionality and have tested it with all combinations I can think of. This uses a modal popup like the Invite User function does, but does not allow changing the email address or user type, just the number of days the invitation is valid so it can be extended. Most of this is in the new RegenerateInvitationForm.tsx file.

- Replaced the method for keeping track of how many invites have been sent per hour to one user in inviteUser.ts. Took the logic out of the initial Invite User functionality. The initial invite will only let you create one invitation for a given email address regardless of time. Moved the new inviteTracker logic to the regeneration of invites. User will be able to create initial invite and then can regenerate up to 3 times within a one hour period. I changed your exiting logic to use node-cache instead of the previous solution where there was a comment of "// TODO: this could cause memory increase over time if the object is never deleted". I think node-cache will solve that, as it should clean up after itself. ~I haven't actually tested to ensure it is successfully reset after an hour, so should test that.~ Confirmed working.

- Changed Invite User and Regenerate Invitation modals so that they always default to send email. Before, if you invited multiple users or regenerated multiple invitations, it would maintain the last state it was in. I think setting the default every time the modal opens and making sure the form is consistent is a cleaner user experience and the default assumption is that an invite is sent and not sending an email would be the exception.

- Changed invitation email subject from "You're invited to join a Fossorial organization" to "You're invited to join {orgId}". I think Fossorial might be confusing to end/less-technical users? Maybe this customization could be behind a supporter key? Didn't change the signature in this commit, but also possibly something to change/enable with supporter key.


Screenshots:
![image](https://github.com/user-attachments/assets/c879a22c-ddf2-4480-9071-cfb61b8ebdeb)
![image](https://github.com/user-attachments/assets/7ad6c748-8030-4669-b0cb-3a0731c85a08)
![screenshot_20250409_20 51 53_32ibIt](https://github.com/user-attachments/assets/ff2caa16-18ae-4518-9a85-ed8cad526c7e)
![screenshot_20250409_20 53 13_xqEtbD](https://github.com/user-attachments/assets/ab1b3632-470a-4ffa-bc5b-e68fd6916897)
![screenshot_20250409_20 50 02_gTZVPf](https://github.com/user-attachments/assets/066a8e12-60a2-4c38-869f-87f260325af7)
![47491](https://github.com/user-attachments/assets/b651b71d-a935-4e7b-9795-86b5e0d9f5b3)

miloschwartz commented:
Thank you!
